### PR TITLE
different additions

### PR DESCRIPTION
**Pull request by roessig**
- various methods for class Event
- several methods for variable bounds
- nodeLowerbound added
- created method delObj that sets all objective coefficients to 0.0
- some LP and probing methods
- branchVar and branchVarVal added, not sure about how /if to release the memory

### CODE REVIEW

File: src/pyscipopt/scip.pyx [hunk 1]
```diff
@@ -766,6 +780,16 @@ cdef class Model:
...
+    def delObjective(self):
+        """clear existing objective function"""
+        cdef SCIP_VAR** _vars
+        cdef int _nvars
+
+        _vars = SCIPgetOrigVars(self._scip)
+        _nvars = SCIPgetNOrigVars(self._scip)
+        for i in range(_nvars):
+            PY_SCIP_CALL(SCIPchgVarObj(self._scip, _vars[i], 0.0))
+
```
Comment (mattmilten) [inline src/pyscipopt/scip.pyx]: Why is this necessary? Can't you just call `setObjective(0)`?

File: src/pyscipopt/scip.pyx [hunk 2]
```diff
@@ -969,6 +993,78 @@ cdef class Model:
...
+        cdef SCIP_Bool infeasible
+        cdef SCIP_Bool fixed
+        PY_SCIP_CALL(SCIPfixVar(self._scip, var.var, val, &infeasible, &fixed))
+        return infeasible, fixed
+
+    def delVar(self, Variable var):
+        """Delete a variable.
+
+        :param var: the variable which shall be deleted
+        :return bool, was deleting succesful
```
Comment (mattmilten) [inline src/pyscipopt/scip.pyx]:
```suggestion
        :return: bool, was deleting succesful
```

File: src/pyscipopt/scip.pyx [hunk 3]
```diff
@@ -969,6 +993,78 @@ cdef class Model:
...
+        """
+        cdef SCIP_Bool deleted
+        PY_SCIP_CALL(SCIPdelVar(self._scip, var.var, &deleted))
+        return deleted
+
+    def tightenVarLb(self, Variable var, lb, force=False):
+        """Tighten the lower bound in preprocessing or current node, if the bound is tighter.
+        :param var: SCIP variable
```
Comment (mattmilten) [inline src/pyscipopt/scip.pyx]:
```suggestion

        :param var: SCIP variable
```

File: src/pyscipopt/scip.pyx [hunk 4]
```diff
@@ -969,6 +993,78 @@ cdef class Model:
...
+        :return: bool, if the bound was tightened
+        """
+        cdef SCIP_Bool infeasible
+        cdef SCIP_Bool tightened
+        PY_SCIP_CALL(SCIPtightenVarLb(self._scip, var.var, lb, force, &infeasible, &tightened))
+        return tightened
+
+
+    def tightenVarUb(self, Variable var, ub, force=False):
+        """Tighten the upper bound in preprocessing or current node, if the bound is tighter.
```
Comment (mattmilten) [inline src/pyscipopt/scip.pyx]:
```suggestion
        """Tighten the upper bound in preprocessing or current node, if the bound is tighter.

```

File: src/pyscipopt/scip.pyx [hunk 5]
```diff
@@ -969,6 +993,78 @@ cdef class Model:
...
+        """
+        cdef SCIP_Bool infeasible
+        cdef SCIP_Bool tightened
+        PY_SCIP_CALL(SCIPtightenVarUb(self._scip, var.var, ub, force, &infeasible, &tightened))
+        return tightened
+    
+    
+    def tightenVarUbGlobal(self, Variable var, ub, force=False):
+        """Tighten the global upper bound, if the bound is tighter.
+        :param var: SCIP variable
```
Comment (mattmilten) [inline src/pyscipopt/scip.pyx]:
```suggestion

        :param var: SCIP variable
```

File: src/pyscipopt/scip.pyx [hunk 6]
```diff
@@ -969,6 +993,78 @@ cdef class Model:
...
+        :return: bool, if the bound was tightened
+        """
+        cdef SCIP_Bool infeasible
+        cdef SCIP_Bool tightened
+        PY_SCIP_CALL(SCIPtightenVarUbGlobal(self._scip, var.var, ub, force, &infeasible, &tightened))
+        return tightened
+    
+    def tightenVarLbGlobal(self, Variable var, lb, force=False):
+        """Tighten the global upper bound, if the bound is tighter.
+        :param var: SCIP variable
```
Comment (mattmilten) [inline src/pyscipopt/scip.pyx]:
```suggestion

        :param var: SCIP variable
```

File: src/pyscipopt/scip.pyx [hunk 7]
```diff
@@ -969,6 +993,78 @@ cdef class Model:
...
+        :param lb: possible new upper bound
+        :param force: force tightening even if below bound strengthening tolerance
+        :return: bool, if the bound was tightened
+        """
```
Comment (mattmilten) [inline src/pyscipopt/scip.pyx]:
```suggestion

        """
```

File: src/pyscipopt/scip.pyx [hunk 8]
```diff
@@ -969,6 +993,78 @@ cdef class Model:
...
+        return tightened
+    
+    
+    def tightenVarUbGlobal(self, Variable var, ub, force=False):
+        """Tighten the global upper bound, if the bound is tighter.
+        :param var: SCIP variable
+        :param ub: possible new upper bound
+        :param force: force tightening even if below bound strengthening tolerance
+        :return: bool, if the bound was tightened
+        """
```
Comment (mattmilten) [inline src/pyscipopt/scip.pyx]:
```suggestion

        """
```

File: src/pyscipopt/scip.pyx [hunk 9]
```diff
@@ -969,6 +993,78 @@ cdef class Model:
...
+        return tightened
+
+
+    def tightenVarUb(self, Variable var, ub, force=False):
+        """Tighten the upper bound in preprocessing or current node, if the bound is tighter.
+        :param var: SCIP variable
+        :param ub: possible new upper bound
+        :param force: force tightening even if below bound strengthening tolerance
+        :return: bool, if the bound was tightened
+        """
```
Comment (mattmilten) [inline src/pyscipopt/scip.pyx]:
```suggestion

        """
```

File: src/pyscipopt/scip.pyx [hunk 10]
```diff
@@ -1030,11 +1171,37 @@ cdef class Model:
 
         return [Variable.create(_vars[i]) for i in range(_nvars)]
 
+    def updateNodeLowerbound(self, Node node, lb):
+        """if given value is larger than the node's lower bound (in transformed problem),
+        sets the node's lower bound to the new value
+        Args:
+            node: Node, the node to update
+            newbound: float, new bound (if greater) for the node
+        """
```
Comment (mattmilten) [inline src/pyscipopt/scip.pyx]: Please use the usual docstring style.

File: src/pyscipopt/scip.pyx [hunk 11]
```diff
@@ -1030,11 +1171,37 @@ cdef class Model:
...
         return SCIPgetLPSolstat(self._scip)
 
+
+    def constructLP(self):
+        """makes sure that the LP of the current node is loaded and
+         may be accessed through the LP information methods
+
+         Returns:
+             cutoff: bool, can the node be cut off?
+         """
```
Comment (mattmilten) [inline src/pyscipopt/scip.pyx]: Please use the usual doc style

File: src/pyscipopt/scip.pyx [hunk 12]
```diff
@@ -2002,7 +2169,7 @@ cdef class Model:
     def getConss(self):
         """Retrieve all constraints."""
         cdef SCIP_CONS** _conss
-        cdef SCIP_CONS* _cons
+        # cdef SCIP_CONS* _cons # is not used
```
Comment (mattmilten) [inline src/pyscipopt/scip.pyx]:
```suggestion
```

File: src/pyscipopt/scip.pyx (hunk 1 above)
Comment (roessig) [reply src/pyscipopt/scip.pyx]: setObjective(0) raises an AssertionError at  assert isinstance(coeffs, Expr) in setObjective. Therefore I created this method. Maybe 0 should be an instance of Expr, I don't know

Comment (mattmilten) [reply src/pyscipopt/scip.pyx]: Oh! I was not aware of that. Then we should rather fix this instead of introducing a new method.

**Comment by roessig** [comment]
I changed the documentation accordingly, added the return of "infeasible" in the tightenVar methods and created tests for (almost) everything.

**Comment by roessig** [comment]
Finally the test works also here. Furthermore, I added the function getLPBranchCands in class Model.

File: tests/test_branch_probing_lp.py [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+
+    def __init__(self, model, cont):
+        self.model = model
+        self.cont = cont
+        self.count = 0
+        self.was_called_val = False
+        self.was_called_int = False
+
+    def branchexeclp(self, allowaddcons):
+        print("in branchexelp")
```
Comment (mattmilten) [inline tests/test_branch_probing_lp.py]:
```suggestion

```

Comment (mattmilten) [reply tests/test_branch_probing_lp.py]: this is a debug print and should be removed

File: src/pyscipopt/scip.pyx [hunk 14]
```diff
@@ -2488,6 +2673,75 @@ cdef class Model:
...
+        cdef SCIP_NODE* upchild = <SCIP_NODE*> malloc(sizeof(SCIP_NODE))
+
+        PY_SCIP_CALL(SCIPbranchVarVal(self._scip, (<Variable>variable).var, value, &downchild, &eqchild, &upchild))
+
+        # TODO should the stuff be freed and how?
+        return Node.create(downchild), Node.create(eqchild), Node.create(upchild)
+
+
+    def createChild(self, nodeselprio, estimate):
+        """Create a child node of the focus node."""
```
Comment (mattmilten) [inline src/pyscipopt/scip.pyx]: The return values are missing

File: src/pyscipopt/scip.pyx [hunk 15]
```diff
@@ -2488,6 +2673,75 @@ cdef class Model:
...
+        cdef SCIP_NODE* eqchild = <SCIP_NODE*> malloc(sizeof(SCIP_NODE))
+        cdef SCIP_NODE* upchild = <SCIP_NODE*> malloc(sizeof(SCIP_NODE))
+
+        PY_SCIP_CALL(SCIPbranchVar(self._scip, (<Variable>variable).var, &downchild, &eqchild, &upchild))
+
+        return Node.create(downchild), Node.create(eqchild), Node.create(upchild)
+
+
+    def branchVarVal(self, variable, value):
+        """Branches on variable using a value which separates the domain of the variable."""
```
Comment (mattmilten) [inline src/pyscipopt/scip.pyx]: Return values are missing

File: src/pyscipopt/scip.pyx [hunk 16]
```diff
@@ -2488,6 +2673,75 @@ cdef class Model:
...
+        PY_SCIP_CALL(SCIPgetLPBranchCands(self._scip, &lpcands, &lpcandssol, &lpcandsfrac,
+                                          &nlpcands, &npriolpcands, &nfracimplvars))
+
+        return ([Variable.create(lpcands[i]) for i in range(ncands)], [lpcandssol[i] for i in range(ncands)],
+                [lpcandsfrac[i] for i in range(ncands)], nlpcands, npriolpcands, nfracimplvars)
+
+
+    def branchVar(self, variable):
+        """Branch on a non-continuous variable."""
+
```
Comment (mattmilten) [inline src/pyscipopt/scip.pyx]: Return values are not documented

File: src/pyscipopt/scip.pyx (hunk 1 above)
Comment (mattmilten) [reply src/pyscipopt/scip.pyx]: This has been fixed now in the master branch. Please remove `delObjective()`.

**Comment by mattmilten** [comment]
Thank you for your contribution! This is almost ready to be merged.
